statshandler.go

```diff
@@ -52,12 +78,101 @@ func (h *statshandler) HandleRPC(ctx context.Context, s stats.RPCStats) {
+				updateStats(h.jctx, v, false)
+				for idx, kv := range v.Kv {
+					updateStatsKV(h.jctx, false, 0)
+					switch kvvalue := kv.Value.(type) {
+					case *na_pb.KeyValue_UintValue:
+						if kv.Key == "__timestamp__" {
+							var re_c_ts uint64 = 0
+							var re_p_get_ts uint64 = 0
+							if len(v.Kv) > idx+2 {
+								nextKV := v.Kv[idx+1]
+								if nextKV.Key == "__junos_re_stream_creation_timestamp__" {
+									re_c_ts = nextKV.GetUintValue()
+								}
+								nextnextKV := v.Kv[idx+2]
+								if nextnextKV.Key == "__junos_re_payload_get_timestamp__" {
+									re_p_get_ts = nextnextKV.GetUintValue()
+								}
+							}
+
+							//"sensor-path", "sequence-number", "component-id", "sub-component-id", "packet-size", "p-ts", "e-ts", "re-stream-creation-ts", "re-payload-get-ts"))
+							h.jctx.config.CsvStatsJtimon.logger.Printf(
+								fmt.Sprintf("%s,%d,%d,%d,%d,%d,%d,%d,%d\n",
+									v.Path, v.SequenceNumber, v.ComponentId, v.SubComponentId, s.(*stats.InPayload).Length, v.Timestamp, kvvalue.UintValue, re_c_ts, re_p_get_ts))
+						}
+					}
+				}
+			case *gnmi_pb.SubscribeResponse:
+				stat := getKPIStats(v)
+				if stat != nil && stat.Timestamp != 0 {
+					path := stat.SensorName + ":" + stat.Streamed_path + ":" + stat.Path + ":" + stat.Component
+					h.jctx.config.CsvStatsJtimon.logger.Printf(
+						fmt.Sprintf("%s,%d,%d,%d,%d,%d,%d,%d,%d\n",
+							path, stat.SequenceNumber, stat.ComponentId, stat.SubComponentId,
+							s.(*stats.InPayload).Length, stat.notif_timestamp, int64(stat.Timestamp*uint64(1000000)),
+							int64(stat.re_stream_creation_timestamp*uint64(1000000)),
+							int64(stat.re_payload_get_timestamp*uint64(1000000)),
+						),
+					)
+				}
+			}
+		}
 	case *stats.InTrailer:
 	case *stats.End:
 	default:
 	}
 }
 
+func getKPIStats(subResponse *gnmi_pb.SubscribeResponse) *kpiStats {
+
+	stats := new(kpiStats)
+	notfn := subResponse.GetUpdate()
+	if notfn == nil {
+		return nil
+	}
+	stats.notif_timestamp = notfn.Timestamp
+	extns := subResponse.GetExtension()
+
+	if extns != nil {
+		extn := extns[0]
+		if extn != nil {
+			var hdr gnmi_juniper_header_ext.GnmiJuniperTelemetryHeaderExtension
+
+			reg_extn := extn.GetRegisteredExt()
+			msg := reg_extn.GetMsg()
+			err := proto.Unmarshal(msg, &hdr)
+			if err != nil {
+				log.Fatal("unmarshaling error: ", err)
+			}
+
+			stats.ComponentId = hdr.ComponentId
+			stats.SequenceNumber = hdr.SequenceNumber
+			stats.Path = hdr.SubscribedPath
+			stats.SubComponentId = hdr.SubComponentId
+			stats.Component = hdr.Component
+			stats.Streamed_path = hdr.StreamedPath
+			stats.SensorName = hdr.SensorName
+
+			if hdr.ExportTimestamp > 0 {
+				stats.Timestamp = uint64(hdr.ExportTimestamp)
+			}
+			if hdr.PayloadGetTimestamp > 0 {
+				stats.re_payload_get_timestamp = uint64(hdr.PayloadGetTimestamp)
+			}
+			if hdr.StreamCreationTimestamp > 0 {
+				stats.re_stream_creation_timestamp = uint64(hdr.StreamCreationTimestamp)
+			}
+		}
+	}
```

Comment:
If we do not find extension, we may probably be parsing packets from older releases. If we do not add this logic, at the least can we log an error stating that we are not able to derive stats from this packet.

Reply:
done

Reply:
No error log added here ?

Reply:
forgot to push the change. Pushed now